sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -91,6 +92,10 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
       SimplifyCasts,
       SimplifyCaseConversionExpressions,
       EliminateSerialization) ::
+    // Because ColumnPruning is called after PushPredicateThroughProject, the predicate push down
+    // is reversed. This batch is to ensure Filter is pushed below Project, if possible.
+    Batch("Push Predicate Through Project", Once,
+      PushPredicateThroughProject) ::
```

Comment:
Put this role in a separate batch is not correct, some other filter push down rules depend on this one.